From 6d48ae579f988a46eea39446f3366d5157b0a4bd Mon Sep 17 00:00:00 2001
From: Maddison Hellstrom <maddy@na.ai>
Date: Mon, 25 May 2020 20:40:56 -0700
Subject: [PATCH] [plugins/spotify] fix broken integration

I recommend not using purely hierarchical css/xpath selectors as they
will be very brittle, breaking on any minor change to the DOM tree.

Additionally, it's better to query/test for attributes/classnames that
aren't meant to contain text to be displayed to the user (e.g. don't use
titles as previously used in getPlaybackStatus), because it's likely
that Spotify's i18n will cause breakage.
---
 src/plugins/spotify/integration.js | 16 ++++++++--------
 1 file changed, 8 insertions(+), 8 deletions(-)

diff --git a/src/plugins/spotify/integration.js b/src/plugins/spotify/integration.js
index 6d195e01..1b4aa916 100644
--- a/src/plugins/spotify/integration.js
+++ b/src/plugins/spotify/integration.js
@@ -23,19 +23,19 @@ function isOnLoginPage() {
 
 function getButtons() {
     function getPlayPauseButton() {
-        return getElementByXpath('/html/body/div[3]/div/div[3]/div[3]/footer/div/div[2]/div/div[1]/div[3]/button');
+        return document.querySelector('.Root__now-playing-bar button.spoticon-pause-16, .Root__now-playing-bar button.spoticon-play-16');
     }
 
     function getSkipPreviousSongButton() {
-        return getElementByXpath('/html/body/div[3]/div/div[3]/div[3]/footer/div/div[2]/div/div[1]/div[2]/button');
+        return document.querySelector('.Root__now-playing-bar button.spoticon-skip-back-16');
     }
 
     function getSkipNextSongButton() {
-        return getElementByXpath('/html/body/div[3]/div/div[3]/div[3]/footer/div/div[2]/div/div[1]/div[4]/button');
+        return document.querySelector('.Root__now-playing-bar button.spoticon-skip-forward-16');
     }
 
     function getAddRemoveToMusicButton() {
-        return getElementByXpath('/html/body/div[3]/div/div[3]/div[3]/footer/div/div[1]/div/div[3]/button');
+        return document.querySelector('.Root__now-playing-bar button.spoticon-heart-16, .Root__now-playing-bar button.spoticon-heart-active-16');
     }
 
     return {
@@ -50,14 +50,14 @@ function getPlaybackStatus() {
     var button = getButtons().playPause;
     if (button === null)
         return MellowPlayer.PlaybackStatus.STOPPED;
-    else if (button.title === "Play")
+    else if (button.classList.contains("spoticon-play-16"))
         return MellowPlayer.PlaybackStatus.PAUSED;
     return MellowPlayer.PlaybackStatus.PLAYING;
 }
 
 function getArtist() {
     try {
-        return getElementByXpath('/html/body/div[3]/div/div[3]/div[3]/footer/div/div[1]/div/div[2]/div[2]').children[0].innerText
+        return document.querySelector('.Root__now-playing-bar .now-playing > div:nth-of-type(2) a[href^="/album/"]').innerText
     } catch (e) {
         return ""
     }
@@ -65,7 +65,7 @@ function getArtist() {
 
 function getSongTitle() {
     try {
-        return getElementByXpath('/html/body/div[3]/div/div[3]/div[3]/footer/div/div[1]/div/div[2]/div[1]/div/span/a').innerText;
+        return document.querySelector('.Root__now-playing-bar .now-playing a[href^="/artist/"]').innerText
     } catch (e) {
         return ""
     }
@@ -131,7 +131,7 @@ function isFavorite() {
     if (addRemoveToMusic === null) {
         return false;
     }
-    return addRemoveToMusic.attributes.class.value.match(".*active.*") !== null;
+    return addRemoveToMusic.classList.contains("spoticon-heart-active-16")
 }
 
 function update() {
-- 
2.26.2

